File: mobile/android/app/src/main/kotlin/com/hearth/backup/DumpV1.kt
package com.hearth.backup

import com.hearth.data.Call
import com.hearth.data.Provider
import com.hearth.data.Repo
import com.hearth.genome.decode
import com.hearth.genome.gunzip
import org.json.JSONObject
import javax.crypto.AEADBadTagException
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Dump v1, the format before dump v2 (export/dump.ts): one gzipped JSON document, optionally after
 * `HRTH1` + 16-byte salt + 12-byte nonce + AES-GCM. Genotypes are a string table over a shared SNP
 * index, two characters per SNP per person. Read only: the app writes v2.
 */
private val V1_MAGIC = "HRTH1".toByteArray()

/** A v1 dump sealed with a passphrase. */
fun isSealedV1(bytes: ByteArray): Boolean = bytes.size > 5 && V1_MAGIC.indices.all { bytes[it] == V1_MAGIC[it] }

fun isV1(bytes: ByteArray): Boolean = isSealedV1(bytes) || (bytes.size > 2 && bytes[0] == 0x1f.toByte() && bytes[1] == 0x8b.toByte())

fun openV1(bytes: ByteArray, passphrase: String?): JSONObject {
    var gz = bytes
    if (isSealedV1(bytes)) {
        if (passphrase.isNullOrEmpty()) throw BackupException("native.encryptedTitle")
        val salt = bytes.copyOfRange(5, 21)
        val nonce = bytes.copyOfRange(21, 33)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(pbkdf2Sha256(passphrase, salt, 600_000), "AES"), GCMParameterSpec(128, nonce))
        gz = try {
            cipher.doFinal(bytes, 33, bytes.size - 33)
        } catch (_: AEADBadTagException) {
            throw BackupException("app.wrongPassphrase")
        }
    }
    val d = try {
        JSONObject(decode(gunzip(gz)))
    } catch (_: Exception) {
        throw BackupException("native.damaged")
    }
    if (d.optString("format") != "hearth-dump" || d.optInt("version") != 1) throw BackupException("native.notBackup")
    return d
}

/** personId → calls, from the shared SNP index (dump.ts `expandDump`); "--" is no call. */
fun expandDump(d: JSONObject): Map<String, List<Call>> {
    val idx = d.getJSONObject("snp_index")
    val rsids = idx.getJSONArray("rsids")
    val chromosomes = idx.getJSONArray("chromosomes")
    val positions = idx.getJSONArray("positions")
    val g = d.getJSONObject("genotypes")
    return g.keys().asSequence().associateWith { pid ->
        val s = g.getString(pid)
        (0 until rsids.length()).mapNotNull { i ->
            val a1 = s[i * 2].toString()
            val a2 = s[i * 2 + 1].toString()
            if (a1 == "-" && a2 == "-") null else Call(rsids.getString(i), chromosomes.getString(i), positions.getLong(i), a1, a2)
        }
    }
}

/**
 * A v1 dump merged in (restore.ts `restoreV1`): people not here yet with their genotypes, the
 * links, their health entries, the consents. People already here are left alone.
 */
fun restoreV1(repo: Repo, d: JSONObject): RestoreResult = repo.sql.transaction {
    val calls = expandDump(d)
    val existing = repo.listPersons().map { it.id }.toSet()
    val added = mutableSetOf<String>()
    val sourceFiles = d.optJSONArray("source_files")
    val persons = d.getJSONArray("persons")
    val healthBefore = repo.listHealthLog().size
    for (i in 0 until persons.length()) {
        val p = persons.getJSONObject(i)
        val id = p.getString("id")
        if (id in existing) continue
        added.add(id)
        repo.sql.exec(
            "INSERT INTO person(id,label,display_name,sex,birth_year,notes,created_at) VALUES (?,?,?,?,?,?,?)",
            listOf(id, p.optString("label"), p.optString("displayName"), p.optString("sex", "unknown"), if (p.isNull("birthYear")) null else p.optInt("birthYear"), p.optString("notes"), p.optString("createdAt")),
        )
        val sf = (0 until (sourceFiles?.length() ?: 0)).map { sourceFiles!!.getJSONObject(it) }.firstOrNull { it.optString("personId") == id }
        repo.importCalls(
            id,
            Provider.of(sf?.optString("provider") ?: "generic"),
            sf?.optString("build")?.ifEmpty { null } ?: "37",
            sf?.optString("sha256") ?: "",
            sf?.optString("originalName")?.ifEmpty { null } ?: "dump-v1",
            calls[id] ?: emptyList(),
        )
    }
    val rels = d.optJSONArray("relationships")
    for (i in 0 until (rels?.length() ?: 0)) rels!!.getJSONObject(i).let { repo.setParent(it.getString("parentId"), it.getString("childId")) }
    val journal = JSONObject()
        .put("health_log", org.json.JSONArray().apply {
            val h = d.optJSONArray("health_log")
            for (i in 0 until (h?.length() ?: 0)) h!!.getJSONObject(i).let { if (it.optString("person_id") in added) put(it) }
        })
        .put("consents", d.optJSONArray("consents") ?: org.json.JSONArray())
    // Health rows and consents go through the v2 merge, with its allowlists and defaults.
    restore(repo, Container(JSONObject().put("exported_at", d.optString("exported_at")), JSONObject(), journal))
    RestoreResult(added.size, repo.listHealthLog().size - healthBefore, added.size, 0, d.optString("exported_at"))
}

File: mobile/ios/Hearth/Native/Backup/DumpV1.swift
import CryptoKit
import Foundation

/// Dump v1, the format before dump v2 (export/dump.ts): one gzipped JSON document, optionally after
/// `HRTH1` + 16-byte salt + 12-byte nonce + AES-GCM. Genotypes are a string table over a shared SNP
/// index, two characters per SNP per person. Read only: the app writes v2.
enum DumpV1 {
    private static let magic = Array("HRTH1".utf8)

    /// A v1 dump sealed with a passphrase.
    static func isSealed(_ data: Data) -> Bool {
        data.count > 5 && Array(data.prefix(5)) == magic
    }

    static func isV1(_ data: Data) -> Bool {
        isSealed(data) || (data.count > 2 && data[data.startIndex] == 0x1f && data[data.startIndex + 1] == 0x8b)
    }

    /// The v1 document, decrypted and inflated.
    static func open(_ data: Data, passphrase: String?) throws -> [String: Any] {
        var gz = data
        if isSealed(data) {
            guard let passphrase, !passphrase.isEmpty else { throw BackupError.passphraseRequired }
            let bytes = [UInt8](data)
            guard bytes.count >= 33 + 16 else { throw BackupError.damaged }
            let key = try Container.deriveKey(passphrase, salt: Array(bytes[5..<21]))
            do {
                let box = try AES.GCM.SealedBox(
                    nonce: AES.GCM.Nonce(data: bytes[21..<33]),
                    ciphertext: bytes[33..<(bytes.count - 16)],
                    tag: bytes[(bytes.count - 16)...]
                )
                gz = try AES.GCM.open(box, using: key)
            } catch {
                throw BackupError.wrongPassphrase
            }
        }
        let document: [String: Any]
        do {
            let text = GenomeImport.decode(try Gzip.decompress(gz))
            guard let d = try JSONSerialization.jsonObject(with: Data(text.utf8)) as? [String: Any] else {
                throw BackupError.damaged
            }
            document = d
        } catch {
            throw BackupError.damaged
        }
        guard document["format"] as? String == "hearth-dump", (document["version"] as? NSNumber)?.intValue == 1 else {
            throw BackupError.notABackup
        }
        return document
    }

    /// personId → calls, from the shared SNP index (dump.ts `expandDump`); "--" is no call.
    static func expand(_ d: [String: Any]) -> [String: [Call]] {
        guard let index = d["snp_index"] as? [String: Any],
              let rsids = index["rsids"] as? [String],
              let chromosomes = index["chromosomes"] as? [String],
              let positions = index["positions"] as? [NSNumber],
              let genotypes = d["genotypes"] as? [String: String]
        else { return [:] }
        var out: [String: [Call]] = [:]
        for (personId, s) in genotypes {
            let chars = Array(s)
            var calls: [Call] = []
            for i in rsids.indices where i * 2 + 1 < chars.count && i < chromosomes.count && i < positions.count {
                let a1 = String(chars[i * 2])
                let a2 = String(chars[i * 2 + 1])
                if a1 == "-" && a2 == "-" { continue }
                calls.append(Call(rsid: rsids[i], chromosome: chromosomes[i], position: positions[i].intValue, a1: a1, a2: a2))
            }
            out[personId] = calls
        }
        return out
    }

    private static func string(_ o: [String: Any], _ key: String, _ fallback: String = "") -> String {
        (o[key] as? String) ?? fallback
    }

    /// A v1 dump merged in (restore.ts `restoreV1`): people not here yet with their genotypes, the
    /// links, their health entries, the consents. People already here are left alone.
    static func restore(_ d: [String: Any], into repo: Repo) throws -> RestoreResult {
        try repo.db.transaction { () throws -> RestoreResult in
            let calls = expand(d)
            let existing = try Set(repo.persons().map(\.id))
            var added = Set<String>()
            let sourceFiles = d["source_files"] as? [[String: Any]] ?? []
            let healthBefore = try repo.healthLog().count
            for p in d["persons"] as? [[String: Any]] ?? [] {
                let id = string(p, "id")
                if id.isEmpty || existing.contains(id) { continue }
                added.insert(id)
                let birthYear = (p["birthYear"] as? NSNumber).map { SQLValue.integer($0.int64Value) } ?? .null
                try repo.db.run(
                    "INSERT INTO person(id,label,display_name,sex,birth_year,notes,created_at) VALUES (?,?,?,?,?,?,?)",
                    [
                        .text(id), .text(string(p, "label")), .text(string(p, "displayName")),
                        .text(string(p, "sex", "unknown")), birthYear, .text(string(p, "notes")), .text(string(p, "createdAt")),
                    ]
                )
                let sf = sourceFiles.first { string($0, "personId") == id }
                try repo.importCalls(
                    personId: id,
                    provider: Provider.of(sf.map { string($0, "provider", "generic") } ?? "generic"),
                    build: sf.map { string($0, "build") }.flatMap { $0.isEmpty ? nil : $0 } ?? "37",
                    sha256: sf.map { string($0, "sha256") } ?? "",
                    originalName: sf.map { string($0, "originalName") }.flatMap { $0.isEmpty ? nil : $0 } ?? "dump-v1",
                    calls: calls[id] ?? []
                )
            }
            for r in d["relationships"] as? [[String: Any]] ?? [] {
                try repo.setParent(parentId: string(r, "parentId"), childId: string(r, "childId"))
            }
            // Health rows and consents go through the v2 merge, with its allowlists and defaults.
            let health = (d["health_log"] as? [[String: Any]] ?? []).filter { added.contains(string($0, "person_id")) }
            let partial: [String: Any] = ["health_log": health, "consents": d["consents"] as? [Any] ?? []]
            let journal = try JSONDecoder().decode(
                [String: JSONValue].self, from: JSONSerialization.data(withJSONObject: partial)
            )
            let exportedAt = string(d, "exported_at")
            let header = DumpHeader(format: "hearth-dump", version: 1, exportedAt: exportedAt, device: nil, generation: nil, encrypted: false)
            _ = try Restore.run(
                Dump(header: header, exportedAt: exportedAt, encrypted: false, journal: journal, genomes: [], genomeFiles: [:]),
                into: repo
            )
            let healthAfter = try repo.healthLog().count
            return RestoreResult(
                persons: added.count, healthEntries: healthAfter - healthBefore, genomes: added.count,
                skippedAttachmentFiles: 0, exportedAt: exportedAt
            )
        }
    }
}
